src/TermInterface.jl

```diff
@@ -106,18 +90,21 @@ function metadata(x, data)
   error("Setting metadata on $x is not possible")
 end
 
-
 """
-  similarterm(x, head, args, symtype=nothing; metadata=nothing, exprhead=:call)
+  similarterm(x, head, args, symtype=nothing; metadata=nothing)
 
 Returns a term that is in the same closure of types as `typeof(x)`,
 with `head` as the head and `args` as the arguments, `type` as the symtype
 and `metadata` as the metadata. By default this will execute `head(args...)`.
-`x` parameter can also be a `Type`. The `exprhead` keyword argument is useful 
-when manipulating `Expr`s.
+`x` parameter can also be a `Type`. Implementers should define similarterm on the
+type of `x` and not on `x` itself.
 """
-function similarterm(x, head, args, symtype = nothing; metadata = nothing, exprhead = nothing)
-    head(args...)
+function similarterm(x, head, args, symtype = nothing; metadata = nothing)
+  similarterm(typeof(x), head, args, symtype = symtype, metadata = metadata)
+end
+
+function similarterm(T::Type, head, args, symtype = nothing; metadata = nothing)
+  head(args...)
```

Comment:
@shashi do you agree on moving `symtype` arg to `type` kwarg (same for function above)? It's more of a naming convention. `symtype` name can be cryptic, and is strictly related to SU, and does not really reflect the idea of "type" of an expression in a typed language.

For example

```
@syms x y 
t = x * y
SymbolicUtils.symtype(t) # Number
```

But in other interpretations (e.g) typed lambda calculi the type of the term may be `(Number * Number) -> Number`

Reply:
This is a really good point. If I was writing C code, what is the `symtype` of `for(int i = 0; i < 10; i++){x += i}`?

I could think of a few answers, I think 1 was intended, but 2 might be more correct:
1. symtype is always the type of the value returned by executing the expression, whatever it is. 
2. symtype is the "type" of the expression, like `variable`, `type`, `statement`, `expression`, etc. This could also extend to encapsulate 1 with type parameters, like `Expression{Float64}`, or `Statement{Nothing}`.
3. symtype is whatever you want it to be have fun

Reply:
One could imagine writing type-checking utilities using option 2, enforcing something like "every for loop needs a variable declaration followed by an arbitrary expression followed by a statement". That said, I think that sort of functionality might not best be suited to live in this package, or if it did it would need to be a separate effort to implement that by someone who feels they need it.

For now, perhaps it would be better to move symtype to terminterface if we can't come up with a good description of what it actually does.

Reply:
Yeah, stuff like `(julia> (@syms x::Int y::Int); t = x * y; typeof(t)` already returns `SymbolicUtils.BasicSymbolic{Int64}` but `SymbolicUtils.symtype` (which is not even exported) returns `Int64`

Reply:
So what do we think, delete symtype?

Reply:
This would be a really great idea to make it a kwarg, except:

If I defined similarterm without a `type`, but some internal code calls it with `type` then it's going to result in a method error.

We could make it a kwarg, but a required kwarg.

Reply:
> But in other interpretations (e.g) typed lambda calculi the type of the term may be (Number * Number) -> Number

That looks like a function type, `x * y` is a value with type `Number`. I think `type` has a clear meaning. It's the type of the S-expression. I'm going to say there's no issue here as it is. Will think about making it a kwarg, but kwargs can lead to lots of ambiguity/missing method issues.

Reply:
You can always fallback to `Any` when you don't want to use it, but again to avoid method mayhem, I'm going to leave it as a required argument for implementors. We want them to define just one method, not 2 just in case someone is not using the optional argument.

Reply:
> This would be a really great idea to make it a kwarg, except:
> 
> If I defined similarterm without a `type`, but some internal code calls it with `type` then it's going to result in a method error.
> 
> We could make it a kwarg, but a required kwarg.

If a package provides `f(a::A; b = nothing)` I think then dependents that implement methods for `f` should always accept `b`. This applies here. If they don't, then they're not implementing the right method.